float32 concentration_1
float32 concentration_2